### ISSUE TITLE
Compatibility with asynchronously resolved handlers

### ISSUE THREAD

**Issue by trentmwillis** [nickiaconis/ember-prefetch#14]
In order to support lazy loading Engines, route handlers can now be present asynchronously. This causes some issues in this addon, since it expects `handlerInfo.handler` to be present synchronously in some of its code.

I believe the only line that is breaking is here: https://github.com/nickiaconis/ember-prefetch/blob/63c96e90a47b12fa0973d5605e9f304e01a78088/addon/instance-initializers/prefetch.js#L41

It seems like the solution here is to follow the synchronous path if the `handler` is available, otherwise we should likely run `prefetch` after `handlerInfo.handlerPromise` has resolved. Thoughts?

**Comment by nathanhammond** [comment]
We should likely flush with whatever handlers we already have, but I'm not sure we have insight into that?

**Comment by nickiaconis** [comment]
The route (i.e. handler) lookup in the `prefetched` method is likely going to cause pain: https://github.com/nickiaconis/ember-prefetch/blob/master/addon/mixins/route.js#L53

**Comment by trentmwillis** [comment]
After discussion in person, this seems relatively straightforward, but we need to guarantee that the `prefetch` hooks fire in the same order as the `handlerInfos` array.

Opened an issue with router.js to make this easy: https://github.com/tildeio/router.js/issues/193

**Comment by nathanhammond** [comment]
@nickiaconis and I continued the conversation on the train ride home. Depending upon the implementation inside of router.js presently this _may_ not matter.

If each of the `handlerInfos` is tied to a single promise for the loading of that engine's assets then:

```
// foo-engine/bar/baz/12

handlerResolves
  // When? When the engine loads?
  // If the handlers are all tied to an identical promise...
  .then()
  // Each of these will resolve synchronously
  .then()
  // In the microtask queue.
```

**Comment by trentmwillis** [comment]
@nathanhammond so that will be more-or-less true for the Engine's implementation, but is not guaranteed to be true by router.js

**Comment by nickiaconis** [comment]
Resolved by #15
